DOC-1191 publish upgrade policy (#267)
redpanda-data/cloud-docs: modules/manage/pages/maintenance.adoc

== Minor upgrades

During your defined maintenance window, Redpanda Cloud runs minor upgrades. Minor upgrades include standard Redpanda state changes that clients handle gracefully, such as leader elections. 

[cols="1,4", options="header"]
|===
| Category
| Details

| Impact
| Minimal.

| Examples
| * Patches to known issues. +
* Cluster rolling restart.  +
* Upgrade Redpanda to a fully backward-compatible version. 

| Frequency
| Frequent: minor upgrades could happen multiple times a day.

| Communication
| Prior communication happens only if necessary. +

There could be email notifications, updated documentation, release notes, or communication from your Redpanda account team.

| Timing
| 100% at Redpanda's discretion during your defined maintenance window.
|===

== Major upgrades

Major upgrades may require code changes to customer applications, such as Kafka clients, and API integrations. 

[cols="1,4", options="header"]
|===
| Category
| Details

| Impact
| Potentially large.

> Major upgrades could require code changes to customer applications or could problematically alter the performance profile. 

Author: fixed

* Security update that materially changes cluster or client throughput.

| Frequency
| Rare.

| Communication
| Email notifications may be sent to registered users with details about the change and available options. +

There could be updated documentation, release notes, and communication from your Redpanda account team.

| Timing
| Major upgrades may be coordinated with customers, but the final date set by Redpanda is not negotiable.
|===

== Deprecations

Deprecations indicate future removal of features that you can currently use. There is no guarantee of equivalent functionality in new versions. Deprecations could be included in major upgrades. 

[cols="1,4", options="header"]
|===
| Category
| Details

| Impact
| Potentially large, if you depend on the feature being deprecated.

| Examples
a|
* Remove a feature from the UI. +
* Shut down an API version. +
* Remove a connector as an option.

| Frequency
| Rare.

| Communication
| Email notifications may be sent to registered users with details about the change and available alternatives. +

There could be updated documentation, release notes, and communication from your Redpanda account team.

| Timing

> | 100% at Redpanda's discretion.

Reviewer: ```suggestion
| At Redpanda's discretion.
```

> Redpanda Cloud may run maintenance operations on any day at any time. You can override this default and schedule a specific maintenance window on your cluster's *Cluster settings* page. 

Reviewer: ```suggestion
Redpanda Cloud may run maintenance operations on any day, at any time. You can override this default and schedule a specific maintenance window on your cluster's *Cluster settings* page. 
```

If you select a *Scheduled* maintenance window, then Redpanda Cloud runs operations on the day and time specified. Maintenance windows typically take six hours. All operations begin during your maintenance window, but some operations may complete after the window closes. All times are in Coordinated Universal Time (UTC).

> TIP: Redpanda Cloud maintenance cycles always start on Tuesdays. Clusters scheduled for maintenance on Tuesdays are updated first, and clusters scheduled on Mondays are updated last. Keep this in mind when sequencing updates for multiple clusters.

Reviewer: Perhaps I'm misunderstanding this. If all maintenance is scheduled to start on Tuesday, why would cluster updates be scheduled for Mondays?

== Minor upgrades

During your defined maintenance window, Redpanda Cloud runs minor upgrades. Minor upgrades include standard Redpanda state changes that clients handle gracefully, such as leader elections. 

[cols="1,3"]

> |===

Reviewer: I'm not sure about the headings in this table. I started reading the content of the Impact column and it wasn't immediately clear how the "headings" were related to each other. Might be just me.

Author: @asimms41: thanks, I've added headers to the tables

| Impact
| Minimal.

| Examples

> | * Patches to known issues. +

Reviewer: Bullet points are not correctly formatted.

* Cluster rolling restart.  +
* Upgrade Redpanda to a fully backward-compatible version. +

| Frequency

> | Frequent: minor upgrades could happen multiple times a day.

Reviewer: ```suggestion
| Minor upgrades could happen multiple times a day.
```

| Communication
| Prior communication happens only if necessary. 

There could be email notifications, updated documentation, release notes, or communication from your Redpanda account team.

| Timing

> | 100% at Redpanda's discretion during your defined maintenance window.

Reviewer: ```suggestion
| At Redpanda's discretion during your defined maintenance window.
```

|===

== Major upgrades

Major upgrades may require code changes to customer applications, such as Kafka clients, and API integrations. 

> [cols="1,3"]

Reviewer: Again, could there be a header row?

|===
| Impact
| Potentially large.

> Major upgrades could require code changes to customer applications or could problematically alter the performance profile. 

Reviewer: ```suggestion
Major upgrades could require code changes to customer applications or could affect the performance profile. 
```

> | Examples

Reviewer: Bullet points are not correctly formatted.

| * Upgrade Kafka to a version that is not fully backward-compatible with the previous version. +
* Update an API version. +
* Security update that materially changes cluster or client throughput.

| Frequency
| Rare.

| Communication
| Email notifications may be sent to registered users with details about the change and available options.

There could be updated documentation, release notes, and communication from your Redpanda account team.

| Timing
| Major upgrades may be coordinated with customers, but the final date set by Redpanda is not negotiable.
|===

== Deprecations

> Deprecations indicate future removal of customer-usable functionality from Redpanda Cloud services. There is no guarantee of equivalent functionality in new versions. Deprecations could be included in major upgrades. 

Reviewer: Not sure about "customer-usable" functionality. Maybe something like "features that you can currently use"?

Author: fixed

> [cols="1,3"]

Reviewer: Same comments as for other tables.

Redpanda Cloud may run maintenance operations on any day at any time. You can override this default and schedule a specific maintenance window on your cluster's *Cluster settings* page. 

If you select a *Scheduled* maintenance window, then Redpanda Cloud runs operations on the day and time specified. Maintenance windows typically take six hours. All operations commence during your maintenance window, but some operations may complete after the window closes. All times are in Coordinated Universal Time (UTC).

TIP: Redpanda Cloud maintenance cycles always start on Tuesdays. Clusters scheduled for maintenance on Tuesdays are updated first, and clusters scheduled on Mondays are updated last. Keep this in mind when sequencing updates for multiple clusters.

== Minor upgrades

Multiple minor upgrades could happen within your defined maintenance window. Minor upgrades include standard Redpanda state changes that clients handle gracefully, such as leader elections. 

[cols="1,3"]
|===
| Impact
| Minimal

| Examples

> | New features +

Reviewer 2: Are we talking about new cloud features here or core features?   Its a rare exception there is a new feature in a minor @piyushredpanda .  not sure if this sholdd be reworded.  

Bug and security fixes is the main Example here.

* Cluster rolling restart  +
* Upgrade Redpanda to a fully backward-compatible version +

| Frequency
| Frequent: minor upgrades could happen multiple times a day.

| Communication
| Prior communication happens only if necessary. 

There could be email notifications, updated documentation, release notes, or communication from your Redpanda account team.

| Timing
| 100% at Redpanda's discretion within your defined maintenance window.
|===

== Major upgrades

Major upgrades may require code changes to customer applications, such as Kafka clients, and API integrations. 

[cols="1,3"]
|===
| Impact
| Large

> A major upgrade could prevent use of the service without code or configuration changes, or it could problematically alter the performance profile. 

Reviewer 3: I would say its rare that an upgrade would prevent the use of a feature without code or configuration changes even for major upgrades i.e. between 24.3.x and 25.1.x. I know @hcoyote suggested three times a year, but I would say its very rare and likely 1 time a year or less.

Author: I changed major upgrades:

- Impact = Potentially large 
- Frequency = Rare

| Examples
| * Upgrade Kafka to a version that is not fully backward-compatible with the previous version +
* Update an API version +
* Security update that materially changes cluster or client throughput

| Frequency
| Approximately three times a year to coincide with major releases of Redpanda, and at least once a year for Kubernetes upgrades.

| Communication
| Email notifications may be sent to registered users with details about the change and available options.

There could be updated documentation, release notes, and communication from your Redpanda account team.

| Timing
| Major upgrades may be coordinated with customers, but the final date set by Redpanda is not negotiable.
|===

== Deprecations

> Deprecations include any removal of customer-usable functionality from Redpanda Cloud services. There is no guarantee of equivalent functionality in new versions. Deprecations could be included in major upgrades. 

Reviewer 2: 'Deprecations indicate impending future removal of"

Deprecation is not removal

TIP: Redpanda Cloud maintenance cycles always start on Tuesdays. Clusters scheduled for maintenance on Tuesdays are updated first, and clusters scheduled on Mondays are updated last. Keep this in mind when sequencing updates for multiple clusters.

== Minor upgrades

> Multiple minor upgrades could happen within the defined maintenance window. Minor upgrades include standard Redpanda state changes that clients handle gracefully, such as leader elections. 

Reviewer 4: Do we document somewhere what Major and Minor means anywhere?

Author: Good question. We say Self-Managed has feature and patch versions [here](https://docs.redpanda.com/redpanda-cloud/get-started/intro-to-events/#redpanda-self-managed-versions) and RPCN has major, minor, and patch versions [here](https://docs.redpanda.com/redpanda-connect/get-started/upgrade/versioning/)—but nothing about Redpanda Cloud versions. We somewhat explain install packs [here](https://docs.redpanda.com/redpanda-cloud/security/cloud-safety-reliability/#packaging)

Reviewer 4: This is probably for another PR and discussion with people, but we have had customers already complaining about the lack of visibility on what we upgrade.

[cols="1,3"]
|===
| Impact
| Minimal

| Examples
| * Cluster rolling restart  +
* Upgrade Redpanda to a fully backward-compatible version +
* Add new features

| Frequency
| Minor upgrades could happen multiple times each day.

| Communication
| Prior communication happens only if necessary. 

> There could be email notifications, updated documentation, release notes, or communication from Sales and Support.

Reviewer 4: Sales, Customer Success, and Support.

Author: reworded as "communication from your Redpanda account team"

| Timing

> | 100% at Redpanda's discretion within the maintenance window.

Reviewer 4: within the customer-defined maintenance window.

|===

== Major upgrades

> Major upgrades may require code changes to customer applications, such as Kafka clients, and API integrations. 

Reviewer 4: This is making me pause a bit because we have shipped in the past backward-incompatible changes in Minor versions. We can only guarantee this if Core commits to it faithfully. I would prefer if @piyushredpanda and/or @mattschumpert give their blessing.

Reviewer 5: Well, shipping backward-incompatible changes in minor is different from having bugs in minors. And we are building software after all, and we strive really really hard to ship as few bugs as possible, but of course there have been, and practically speaking, will continue to be bugs in minors. What we can do, and are doing, is learn from them, and minimize them as much as possible (like the internal library upgrade in SR we did).

No minor upgrades come to my mind where an externally visible incompatible change was delivered deliberately.

Reviewer 2: +1

[cols="1,3"]
|===
| Impact
| Large

A major upgrade could prevent use of the service without code or configuration changes, or it could problematically alter the performance profile. 

| Examples
| * Upgrade Kafka to a version that is not fully backward-compatible with the previous version +
* Update an API version +
* Security update that materially changes cluster or client throughput

| Frequency
| Approximately three times per year to coincide with major releases of Redpanda, and at least once per year for Kubernetes upgrades.

| Communication
| Email notifications may be sent to registered users with details about the change and available options.

> 

Author: no, this was edited to remove the timeline

There could be updated documentation, release notes, and communication from Sales and Support.

| Timing
| Major upgrades may be coordinated with customers, but the final date set by Redpanda is not negotiable.
|===

== Deprecations

Deprecations include any removal of customer-usable functionality from Redpanda Cloud services. There is no guarantee of equivalent functionality in new versions. Deprecations could be included in major upgrades. 

[cols="1,3"]
|===
| Impact
| Potentially large

If you depend on this service, the deprecation could prevent use of service.

| Examples
| * Remove a feature from the UI +
  * Shut down an API version +
  * Remove a connector as an option

| Frequency
| Rare

| Communication
| Email notifications may be sent to registered users with details about the change and available alternatives.

> 

Author: no, this was edited to remove the timeline

TIP: Redpanda Cloud maintenance cycles always start on Tuesdays. Clusters scheduled for maintenance on Tuesdays are updated first, and clusters scheduled on Mondays are updated last. Keep this in mind when sequencing updates for multiple clusters.

== Minor upgrades

During your maintenance window, Redpanda Cloud may run minor upgrades. Minor upgrades include standard Redpanda state changes that clients handle gracefully, such as leader elections. 

[cols="1,3"]
|===
| Impact
| Minimal

| Examples

> | * Redpanda cluster roll +

Reviewer 6: Have we explained anywhere yet what this is?

Reviewer 7: We can change this to "cluster rolling restart" maybe?

Author: fixed

* Upgrade Redpanda to a fully backward-compatible version +

> * Add new features

Reviewer 6: Not sure if we mean for this to be a bulleted list. If so, maybe the coderabbit suggestion below regarding the list items in a single cell will help?

| Frequency

> | Minor upgrades could happen multiple times each day.

Reviewer 7: Let's update this to something like:

"Multiple minor upgrades could happen within the defined maintenance window"

Author: fixed

| Communication
| Prior communication happens only if necessary. 

There could be email notifications, updated documentation, release notes, or communication from Sales and Support.

| Timing
| 100% at Redpanda's discretion within the maintenance window.

> |===

Author: If this renders properly as is in the same cell, is there a need to refactor with your suggestion?

== Major upgrades

> Major upgrades require code changes to customer applications, such as Kafka clients, and API integrations.

Reviewer 7: this should be "may require". 

is it worth it to also note that this may be required to take advantage of new features in Redpanda? Not sure if we also need to mention backwards compatibility here. Generally, we don't need to have customers touch code.

[cols="1,3"]
|===
| Impact
| Large

A major upgrade could prevent use of the service without code or configuration changes, or it could problematically alter the performance profile.

| Examples
| * Upgrade Kafka to a version that is not fully backward-compatible with the previous version +
* Update an API version +
* Security update that materially changes cluster or client throughput

| Frequency
| Approximately three times per year to coincide with major releases of Redpanda, and at least once per year for Kubernetes upgrades.

| Communication
| Guaranteed

> Multiple email notifications are sent to all registered users—starting at least 180 days before the upgrade date—with details about the change and available options.

Reviewer 7: just to call out, we don't actually do this, I think it's a vestige of using another doc as a template for the wording.

Right now, we don't send notifications at all for any upgrade of Redpanda. We *might* do it for k8s upgrades, but I forget

Author: Thank you: I've reworded and removed "guaranteed"

There could be updated documentation, release notes, and communication from Sales and Support.

| Timing
| Major upgrades may be coordinated with customers, but the final date set by Redpanda is not negotiable.
|===

== Deprecations

Deprecations include any removal of customer-usable functionality from Redpanda Cloud services. There is no guarantee of equivalent functionality in new versions. Deprecations could be included in major upgrades. 

[cols="1,3"]
|===
| Impact
| Potentially large

If you depend on this service, the deprecation could prevent use of service.

| Examples
| * Remove a feature from the UI +
  * Shut down an API version +

>   * Remove a connector as an option

Reviewer 6: Is this Kafka Connect only?

Reviewer 7: I think we have some beta apis that have been long lived where we may have done this. KC is probably going to be the big deprecation that will be coming. I can't think of a case where we've pulled any other feature before

Author: @hcoyote: In Cloud, we've said that we've deprecated [PSC v1](https://docs.redpanda.com/redpanda-cloud/get-started/whats-new-cloud/#improved-private-service-connect-support-with-az-affinity). And we haven't said that Serverless Standard was deprecated, but it basically was.

Reviewer 2: I think we need a section that says 'Deprecation' and a section that says 'Feature/API Removal' explaining that features and APIs/ API versions may be no longer supported following a deprecation notice of at least 6 months (or something similar)

Author: @mattschumpert we have [this deprecation policy for Cloud API](https://docs.redpanda.com/redpanda-cloud/manage/api/cloud-api-deprecation-policy/). Is it correct that this policy pertains to all of our APIs? And features?

| Frequency
| Rare

| Communication
| Guaranteed

> Multiple email notifications are sent to all registered users—starting at least 180 days before the deprecation date—with details about the change and available alternatives.

Reviewer 7: just to note, we don't have a formal deprecation policy (which should include guidance on notification timelines). 

I wonder if we should drop the wording on a timeline for notifications from this and state that deprecations will have sufficient pre-notification time or something. Need PM input here.

Author: @david-yu, @towfiqa please confirm the Communication row in [Major upgrades and Deprecations](https://deploy-preview-267--rp-cloud.netlify.app/redpanda-cloud/manage/maintenance/#major-upgrades)!
